### ISSUE TITLE
Implement exceptions

### ISSUE THREAD

**Issue by plajjan** [actonlang/acton#1167]
TODO:
- [ ] add tests
  - [ ] what cases to test?

**Comment by plajjan** [comment]
You @ddjoh mentioned that you had to implement your own conversion function (might it have been str to int?) because the builtin one throws an exception and since we haven't properly implemented exceptions, you weren't able to catch it. This is pretty bad and in a later conversation with @nordlander we decided that it is better to pause his current work on actor / class unification, which is a larger piece of work, and fast-track the exception handling work, at least getting the basics of catching exceptions in there.

@ddjoh you have also mentioned tagged unions quite a lot when we talk about error handling. Do you have some example cases where you think tagged unions would be particularly suitable rather than exceptions or vice versa?

**Comment by nordlander** [comment]
I think tagged unions are intimately connected to the concept of pattern-matching, and Python's recent 'match' statement provides pattern-matching in a rather appealing way. So implementing 'match' (with type inference) is indeed a natural step for us to take rather soon.

With pattern-matching available, a tagged union is essentially just a class with some sub-classes defined, where each sub-class represents an alternative way of constructing a value of the parent type. However, I'd argue that good tagged union support also requires a syntactic ability to freeze such a class hierarchy, so that further alternatives can't be silently added long after supposedly complete pattern-matching functions are defined. I actually have a proposal to that effect:
```
class MyUnion:
    case AltOne:
        x : int
    case AltTwo:
        pass
    case Variant3:
        x : str
        y : bool
```
This would directly correspond to the regular hierarchy
```
class MyUnion:
    pass
class AltOne from MyUnion:
    x : int
class AltTwo from MyUnion:
    pass
class Variant3 from MyUnion:
    x : str
    y : bool
```
but with the added constraint that further sub-classes can't be added to `MyUnion` if defined using my proposed syntax. I also believe this syntax blends well with other features of our class declarations, like generics and the new constructor parameters, but the details have to be worked out, of course.

Tagged unions to the rescue!

**Comment by ddjoh** [comment]
One example of tagged unions for error handling is what is used by `Rust` which translated into the above could look something like:

```
class Result[R, E]:
    case Ok:
        result : R
    case Err:
        error : E
```

Now instead of something like:

```
class Stream:
    def read_int(self) -> int:
        if (...):
            raise StreamError()
        ...
        return some_int
```

We can do:

```
class Stream:
    def read_int(self) -> Result[int, StreamError]:
        if (...):
            return Err(StreamError())
        ...
        return some_int
```

Chaining multiple matches can become quite verbose, where (again) `Rust` provides an error propagation operator `?` that has special knowledge of `Result` and attempts to match an unwrap the value in Ok(result) or performs a return of Err(error) from the encompassing function.

E.g:

```
def ReadPair(stream: Stream) -> Result[(int, int), StreamError]:
    x = stream.read_int()?
    y = stream.read_int()?
    return (x, y)
```

as equivalent to:


```
def ReadPair(stream: Stream) -> Result[(int, int), StreamError]:
    x_r: Result[int, StreamError] = stream.read_int()
    match x_r:
        case Ok(x):
            y_r: Result[int, StreamError] = stream.read_int()
            match y_r:
                case Ok(y):
                    return (x, y)
                case Err(e):
                    return Err(e)
        case Err(e):
            return Err(e)
```

Corresponding error handling with static type checking can of course also be achieved through checked exceptions which might look more pythonic.

**Comment by plajjan** [comment]
I think this is mostly done now although we are missing tests....

**Comment by plajjan** [comment]
@nordlander do you want to comment on what cases to test for?

**Comment by nordlander** [comment]
I'd argue we're not there yet, the limitation that bans exception handling in **mut** context is probably going to hit the acton-streamer quite often (from what I've seen). I now have a compilation technique for the non-CPS case that's going to work wonders, but while implementing that I've discovered two bugs in the existing CPS-based code as well (which just illustrates that my previous test cases have been insufficient).

So my action plan is to fix the current bugs, implement my new compilation scheme for the non-CPS case, and write a comrehensive suite of unit tests for each combination of
1. CPS/non-CPS contexts
2. Flat/nested try statements
3. Absence/presence of **else** and **final** clauses
4. Result of each clause (fall-through, **return**, **break**, **continue**, and **raise**)

**Comment by plajjan** [comment]
Fixed by #1463. Closing
